packages/xdl/src/createWebpackCompiler.js

```diff
@@ -0,0 +1,168 @@
+  _log(` `);
+}
+
+export function printPreviewNotice(projectRoot) {
+  log(projectRoot, ` `);
+  log(
+    projectRoot,
+    chalk.underline.yellow('Web support in Expo is experimental and subject to breaking changes.')
+  );
+  log(projectRoot, chalk.underline.yellow('Do not use this in production yet.'));
+}
+
+export default function createWebpackCompiler({
+  projectRoot,
+  appName,
+  config,
+  urls,
+  nonInteractive,
+  useYarn,
+  webpack,
+  onFinished,
+}) {
+  const devSocket = {
+    warnings: warnings => logWarning(projectRoot, warnings),
+    errors: errors => logError(projectRoot, errors),
+  };
+
+  // "Compiler" is a low-level interface to Webpack.
+  // It lets us listen to some events and provide our own custom messages.
+  let compiler;
+  try {
+    compiler = webpack(config);
+  } catch (err) {
+    logError(projectRoot, ' ');
+    logError(projectRoot, 'Failed to compile');
+    logError(projectRoot, ' ');
+    logError(projectRoot, err.message || err);
+    logError(projectRoot, ' ');
+    process.exit(1);
+  }
+
+  // "invalid" event fires when you have changed a file, and Webpack is
+  // recompiling a bundle. WebpackDevServer takes care to pause serving the
+  // bundle, so if you refresh, it'll wait instead of serving the old one.
+  // "invalid" is short for "bundle invalidated", it doesn't imply any errors.
+  compiler.hooks.invalid.tap('invalid', () => {
+    if (!nonInteractive) {
+      clearConsole();
+    }
+    log(projectRoot, chalk.white('\nCompiling...'));
+  });
+
+  let isFirstCompile = true;
+
+  // "done" event fires when Webpack has finished recompiling the bundle.
+  // Whether or not you have warnings or errors, you will get this event.
+  compiler.hooks.done.tap('done', async stats => {
+    if (!nonInteractive) {
+      clearConsole();
+    }
+
+    // We have switched off the default Webpack output in WebpackDevServer
+    // options so we are going to "massage" the warnings and errors and present
+    // them in a readable focused way.
+    // We only construct the warnings and errors for speed:
+    // https://github.com/facebook/create-react-app/issues/4492#issuecomment-421959548
+    const statsData = stats.toJson({
+      all: false,
+      warnings: true,
+      errors: true,
+    });
+
+    const messages = formatWebpackMessages(statsData);
+
+    if (messages.errors.length > 0) {
+      devSocket.errors(messages.errors);
+    } else if (messages.warnings.length > 0) {
+      devSocket.warnings(messages.warnings);
+    }
+
+    // New line after the bundle analyzer finishes
+    log(projectRoot, ` `);
+
+    const isSuccessful = !messages.errors.length && !messages.warnings.length;
+    if (isSuccessful) {
+      log(projectRoot, chalk.bold.cyan(`Compiled successfully!`));
+      printPreviewNotice(projectRoot);
+    }
+    if (isSuccessful && (!nonInteractive || isFirstCompile)) {
+      printInstructions(projectRoot, appName, urls, useYarn);
```

Comment:
Right now this message pops up every time I edit a file. I wonder if that's the best way to do this. Especially since the message is rather long, it's hard to see the "Compiled successfully!" message (the most relevant thing at that moment) from all the other stuff we print:
<img width="682" alt="Screenshot 2019-04-11 at 12 02 34" src="https://user-images.githubusercontent.com/497214/55944641-b7f67700-5c51-11e9-80a6-629b6e2c82c2.png">

Reply:
I tried to keep this looking like CRA, but may need to change it